services/motion/builtin/testutils.go

```diff
@@ -0,0 +1,542 @@
+		)
+		test.That(t, err, test.ShouldBeNil)
+		motorMap[name] = thisMotor
+	}
+	return motorMap
+}
+
+type noisyMovementSensor struct {
+	movementsensor.MovementSensor
+	mu         sync.Mutex
+	noise      spatialmath.Pose
+	queryCount int // Apply noise every other iteration when this is > 1
+}
+
+func (m *noisyMovementSensor) Position(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.queryCount++
+	if m.queryCount <= 10 || m.queryCount%2 == 0 {
+		return m.MovementSensor.Position(ctx, extra)
+	}
+	pos, alt, err := m.MovementSensor.Position(ctx, extra)
+	if err != nil {
+		return nil, 0, err
+	}
+	newPos := spatialmath.PoseToGeoPose(spatialmath.NewGeoPose(pos, 0), m.noise)
+	return newPos.Location(), alt, nil
+}
+
+func (m *noisyMovementSensor) CompassHeading(ctx context.Context, extra map[string]interface{}) (float64, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.queryCount <= 10 || m.queryCount%2 == 0 {
+		return m.MovementSensor.CompassHeading(ctx, extra)
+	}
+	heading, err := m.MovementSensor.CompassHeading(ctx, extra)
+	if err != nil {
+		return 0, err
+	}
+	newPos := spatialmath.PoseToGeoPose(spatialmath.NewGeoPose(geo.NewPoint(0, 0), heading), m.noise)
+	return math.Mod(newPos.Heading(), 360), nil
+}
+
+func createMovementSensor(
+	t *testing.T,
+	ctx context.Context,
+	logger logging.Logger,
+	motors map[string]motor.Motor,
+	fakeBase base.Base,
+	origin *geo.Point,
+	noise spatialmath.Pose,
+) movementsensor.MovementSensor {
+	t.Helper()
+	reg, ok := resource.LookupRegistration(movementsensor.API, wheeledodometry.Model)
+	test.That(t, ok, test.ShouldBeTrue)
+
+	odoConf := &wheeledodometry.Config{
+		LeftMotors:        []string{leftMotorName},
+		RightMotors:       []string{rightMotorName},
+		Base:              fakeBase.Name().ShortName(),
+		TimeIntervalMSecs: float64(updateRate),
+	}
+
+	moveConf := resource.Config{
+		Name:                moveSensorName,
+		API:                 movementsensor.API,
+		ConvertedAttributes: odoConf,
+	}
+	deps := resource.Dependencies{}
+	deps[motors[leftMotorName].Name()] = motors[leftMotorName]
+	deps[motors[rightMotorName].Name()] = motors[rightMotorName]
+	deps[fakeBase.Name()] = fakeBase
+
+	sens, err := reg.Constructor(ctx, deps, moveConf, logger)
+	test.That(t, err, test.ShouldBeNil)
+
+	_, err = sens.DoCommand(ctx, map[string]interface{}{"use_compass": true})
+	test.That(t, err, test.ShouldBeNil)
+
+	_, err = sens.DoCommand(ctx, map[string]interface{}{"reset": true, "setLong": origin.Lng(), "setLat": origin.Lat()})
+	test.That(t, err, test.ShouldBeNil)
+	if noise != nil {
+		sens = &noisyMovementSensor{
+			MovementSensor: sens.(movementsensor.MovementSensor),
+			noise:          noise,
+			queryCount:     0,
+		}
+	}
+
+	return sens.(movementsensor.MovementSensor)
+}
+
+func createMockSlamService(name, pcdPath string, origin spatialmath.Pose, move movementsensor.MovementSensor) (*inject.SLAMService, error) {
```

Comment:
its confusing to have both `createInjectedSlam` and then this function. Why do we need both?

Reply:
The above `createInjectedSlam` is a minimal stub that has an injected SLAM struct that fills the interface but always returns a zero pose, for trivial SLAM testing.

This function creates an entire SLAM service that actually functions and returns a pose based on the output of a movement sensor.